### ISSUE TITLE
Add LR schedulers to more algorithms

### ISSUE THREAD

**Issue by alexnikulkov** [aai-institute/tianshou#586]
- [X] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [X] new feature request
- [X] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [X] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, gym, torch, numpy, sys
  print(tianshou.__version__, gym.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```
I see that PG and algorithms which inherit from it support LR schedulers. Can we add support for LR schedulers to more algorithms? I'm primarily interested in DDPG and SAC.

As a more ambitious refactor, can we add support for LR schedulers to all algorithms by allowing a tuple of (optim, LR_scheduler) to be used instead of an optim when we want to use a scheduler?

**Comment by Trinkle23897** [comment]
> Can we add support for LR schedulers to more algorithms? I'm primarily interested in DDPG and SAC.

How about adding it in BasePolicy? `lr_scheduler` will be stored in BasePolicy, and when calling policy.update(), it checks if `lr_scheduler` is not None then `self.lr_scheduler.step()` at the end.

> by allowing a tuple of (optim, LR_scheduler)

I don't think it's a good solution. (1) it needs to pass extra number of arguments; (2) if using tuple, it causes a severe issue of compatibility; (3) you can always overwrite one scheduler that the step function calls other scheduler's step functions, and it's cleaner to solve this issue.

**Comment by alexnikulkov** [comment]
The scheduler has to be optimizer-specific, so I'm not sure how we could add a single lr_scheduler to BasePolicy. Different algorithms have different numbers and identities of optimizers

>  (1) it needs to pass extra number of arguments;

No extra arguments would be needed. The scheduler would be part of the existing optimizer argument

> (2) if using tuple, it causes a severe issue of compatibility;

It would be backwards compatible with the current version (except for PG-based algorithms, which currently implement their own way of scheduling). If we replace `Optimizer` with `Union[Optimizer, Tuple[Optimizer, LRScheduler]]`, passing in just an optimizer will remain a valid call

> (3) you can always overwrite one scheduler that the step function calls other scheduler's step functions, and it's cleaner to solve this issue.

Can you explain this in more detail?

**Comment by Trinkle23897** [comment]
> Can you explain this in more detail?

```python
class MyScheduler(torch.optim.Optimizer):
  def __init__(self, *args: List[LambdaLR]):
    self.schedulers = args
  def step(self):
    for scheduler in self.schedulers:
      scheduler.step()
```

And passing this to BasePolicy.

**Comment by alexnikulkov** [comment]
Makes sense to me.
[EDIT: removed my wrong suggestion to use  `torch.optim.lr_scheduler.ChainedScheduler` (it only works if the schedulers are all for the same optimizer)]

**Comment by Trinkle23897** [comment]
Would you like to open a pull request?

**Comment by alexnikulkov** [comment]
OK, I can take a look at this when I have some time
